core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+public class Context {
+
+  /**
+   * Stack of context objects which is used to record attach & detach history on a thread.
+   */
+  private static final ThreadLocal<ArrayDeque<Context>> contextStack =
```

Comment:
Thanks for the pointer Eirik, definitely interested in getting more opinion on this..
looping in @adriancole 

So context as defined here is approximately equivalent to ContinuationLocal as defined by the impl you pointed me to. The obvious differences being...

1. The explicit use of string keys for bound values (mostly for debuggability)
2. Immutability after creation and override by occlusion when inheriting
3. We don't assume cross-thread inheritance, we provide utility for it
4  Cancellation which we have an explicit need for.

Zipkin/Dapper is a use-case we clearly have in mind for this so feedback definitely welcome. We've been talking to the Dapper folks about this quite a bit.